ext/rugged/rugged_rebase.c

```diff
@@ -0,0 +1,263 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "rugged.h"
+
+extern VALUE rb_mRugged;
+extern VALUE rb_cRuggedTagAnnotation;
+extern VALUE rb_cRuggedTree;
+extern VALUE rb_cRuggedCommit;
+extern VALUE rb_cRuggedBlob;
+extern VALUE rb_cRuggedRepo;
+
+VALUE rb_cRuggedRebase;
+VALUE rb_cRuggedRebaseOperation;
+
+VALUE ruged_rebase_operation_new(git_rebase_operation *operation, VALUE owner);
+
+static void parse_rebase_options(git_rebase_options *ret, VALUE rb_options)
+{
+	VALUE val;
+
+	if (NIL_P(rb_options))
+		return;
+
+	val = rb_hash_aref(rb_options, CSTR2SYM("quiet"));
+	if (RTEST(val))
+		ret->quiet = 1;
+
+	val = rb_hash_aref(rb_options, CSTR2SYM("rewrite_notes_ref"));
+	if (!NIL_P(val)) {
+		Check_Type(val, T_STRING);
+		ret->rewrite_notes_ref = StringValueCStr(val);
+	}
+
+	rugged_parse_checkout_options(&ret->checkout_options, rb_options);
+}
+
+void rb_git_rebase__free(git_rebase *rebase)
+{
+	git_rebase_free(rebase);
+}
+
+VALUE rugged_rebase_new(VALUE klass, VALUE owner, git_rebase *rebase)
+{
+	VALUE rb_rebase = Data_Wrap_Struct(klass, NULL, &rb_git_rebase__free, rebase);
+	rugged_set_owner(rb_rebase, owner);
+	return rb_rebase;
+}
+
+static VALUE rb_git_rebase_new(int argc, VALUE* argv, VALUE klass)
+{
+	int error;
+	git_rebase *rebase;
+	git_repository *repo;
+	git_annotated_commit *branch = NULL, *upstream = NULL, *onto = NULL;
+	git_oid id;
+	VALUE rb_repo, rb_branch, rb_upstream, rb_onto, rb_options;
+	git_rebase_options options = GIT_REBASE_OPTIONS_INIT;
+
+	rb_scan_args(argc, argv, "31:", &rb_repo, &rb_branch, &rb_upstream, &rb_onto, &rb_options);
+	Data_Get_Struct(rb_repo, git_repository, repo);
+	Check_Type(rb_branch, T_STRING);
+	Check_Type(rb_upstream, T_STRING);
+
+	if ((error = git_oid_fromstr(&id, StringValueCStr(rb_branch))) < 0 ||
+	    (error = git_annotated_commit_lookup(&branch, repo, &id)) < 0 ||
+	    (error = git_oid_fromstr(&id, StringValueCStr(rb_upstream))) < 0 ||
+	    (error = git_annotated_commit_lookup(&upstream, repo, &id)) < 0)
+		goto cleanup;
+
+	if (!NIL_P(rb_onto)) {
+		Check_Type(rb_onto, T_STRING);
+		if ((error = git_oid_fromstr(&id, StringValueCStr(rb_onto))) < 0 ||
+		    (error = git_annotated_commit_lookup(&onto, repo, &id)) < 0)
+			goto cleanup;
+	}
+
+	parse_rebase_options(&options, rb_options);
```

Comment:
You need to move this in front of the other libgit2 calls as well, as `parse_rebase_options` can potentially raise a ruby exception.